flink-mesos/src/main/scala/org/apache/flink/mesos/scheduler/TaskMonitor.scala

```diff
@@ -0,0 +1,258 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.mesos.scheduler
+
+import grizzled.slf4j.Logger
+
+import akka.actor.{Actor, FSM, Props}
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.mesos.scheduler.ReconciliationCoordinator.Reconcile
+import org.apache.flink.mesos.scheduler.TaskMonitor._
+import org.apache.flink.mesos.scheduler.messages.{Connected, Disconnected, StatusUpdate}
+import org.apache.mesos.Protos.TaskState._
+import org.apache.mesos.{SchedulerDriver, Protos}
+
+import scala.PartialFunction.empty
+import scala.concurrent.duration._
+
+/**
+  * Monitors a Mesos task throughout its lifecycle.
+  *
+  * Models a task with a state machine reflecting the perceived state of the task in Mesos.   The state
+  * is primarily updated when task status information arrives from Mesos.
+  *
+  * The associated state data primarily tracks the task's goal (intended) state, as persisted by the scheduler.
+  * Keep in mind that goal state is persisted before actions are taken.    The goal state strictly transitions
+  * thru New->Launched->Released.
+  *
+  * Unlike most exchanges with Mesos, task status is delivered at-least-once, so status handling should be idempotent.
+  */
+class TaskMonitor(
+    flinkConfig: Configuration,
+    schedulerDriver: SchedulerDriver,
+    goalState: TaskGoalState) extends Actor with FSM[TaskMonitorState,StateData] {
+
+  val LOG = Logger(getClass)
+
+  startWith(Suspended, StateData(goalState))
+
+  // ------------------------------------------------------------------------
+  //  Suspended State
+  // ------------------------------------------------------------------------
+
+  when(Suspended) {
+    case Event(update: TaskGoalStateUpdated, _) =>
+      stay() using StateData(update.state)
+    case Event(msg: StatusUpdate, _) =>
+      stay()
+    case Event(msg: Connected, StateData(goal: New)) =>
+      goto(New)
+    case Event(msg: Connected, StateData(goal: Launched)) =>
+      goto(Reconciling)
+    case Event(msg: Connected, StateData(goal: Released)) =>
+      goto(Killing)
+  }
+
+  // ------------------------------------------------------------------------
+  //  New State
+  // ------------------------------------------------------------------------
+
+  when(New) {
+    case Event(TaskGoalStateUpdated(goal: Launched), _) =>
+      goto(Staging) using StateData(goal)
+  }
+
+  // ------------------------------------------------------------------------
+  //  Reconciliation State
+  // ------------------------------------------------------------------------
+
+  onTransition {
+    case _ -> Reconciling =>
+      nextStateData.goal match {
+        case goal: Launched =>
+          val taskStatus = Protos.TaskStatus.newBuilder()
+            .setTaskId(goal.taskID).setSlaveId(goal.slaveID).setState(TASK_STAGING).build()
+          context.parent ! Reconcile(Seq(taskStatus))
```

Comment:
I agree that the alternative to using the implicit parent reference is to always use an explicit reference.     It can be unit tested either way.      I actually considered the latter but couldn't think of a nice name for the explicit reference.